Warn about a weak API-Key at startup

jkoenig134 opened the pull request:
# Readiness checklist

-   [ ] I added/updated tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

# Description

Additionally it is not possible anymore to start the connector without an API Key in production mode.

jkoenig134 commented:
Hmm, some downside - the warning is now spammed in the tests. Maybe it should be disabled when tests are detected.

Milena-Czierlinski commented on src/infrastructure/httpServer/HttpServer.ts:
```diff
@@ -203,6 +190,40 @@ export class HttpServer extends ConnectorInfrastructure<HttpServerConfiguration>
+        ]);
+        if (apiKeyStrength.value !== "Strong") {
+            this.logger.warn(
+                "The configured API key does not meet the requirements. It must be 30 chars and contain at least 1 digit, 1 uppercase, 1 lowercase, and 1 special char."
```
It must be [at least] 30 chars [long]?

Milena-Czierlinski commented on src/infrastructure/httpServer/HttpServer.ts:
```diff
@@ -203,6 +190,40 @@ export class HttpServer extends ConnectorInfrastructure<HttpServerConfiguration>
+        this.app.use(async (req, res, next) => {
+            const apiKeyFromHeader = req.headers["x-api-key"];
+            if (!apiKeyFromHeader || apiKeyFromHeader !== this.configuration.apiKey) {
+                // sleep for 1 to 4 seconds
+                await sleep(1000 * (Math.floor(Math.random() * 4) + 1));
```
Why do we do this?

jkoenig134 replied:
To make brute-froce way harder.

jkoenig134 replied:
I changed it from "one second" to "frome one to four second" to make it way harder for an attacker to estimate when the timeout hits.

Milena-Czierlinski commented:
> the warning is now spammed in the tests

Why don't we have a strong API-Key? 🧐

jkoenig134 commented:
@Milena-Czierlinski also an Option, let me think about this.

britsta commented on src/infrastructure/httpServer/HttpServer.ts:
```diff
@@ -203,6 +190,40 @@ export class HttpServer extends ConnectorInfrastructure<HttpServerConfiguration>
+        this.app.use(async (req, res, next) => {
+            const apiKeyFromHeader = req.headers["x-api-key"];
+            if (!apiKeyFromHeader || apiKeyFromHeader !== this.configuration.apiKey) {
+                // sleep for 1 to 4 seconds
```
Remove comment please.

jkoenig134 replied:
I have added this on purpose b/c I was not sure if the 1-4 seconds logic is understandable without. I'm fine with removing it if everyone understands it.

britsta replied:
I think this line is easily looked-up in case someone is interested in the time of sleep. IMO our guideline to not use comments is more important here.

britsta commented on src/infrastructure/httpServer/HttpServer.ts:
```diff
@@ -203,6 +190,40 @@ export class HttpServer extends ConnectorInfrastructure<HttpServerConfiguration>
+            }
+        }
+
+        const apiKeyStrength = checkAPIKeyStrength(this.configuration.apiKey, [
+            { id: 0, value: "Very Weak", minDiversity: 0, minLength: 0 },
+            { id: 1, value: "Strong", minDiversity: 0, minLength: 30 }
+        ]);
```
Do I get it right, that a password of 29 characters and a lot of digitis, special characters and uppercase and lowercase letters is now displayed as "Very Weak"? I get that it does not meet our requirements, but maybe we should call it a "Weak" password then if we don't want to have a finer subdivision of password strengths.

jkoenig134 replied:
I have to define a "low bound", the name here is irrelevant.

stnmtz commented:
I'm absolutely no fan of introducing an external password-check library for such a low-level check. I would recommend removing this external dependency for security reasons and match the following regexp which still needs to be fit to our requirements (see https://stackoverflow.com/questions/5142103/regex-to-validate-password-strength).

> ^                         Start anchor
(?=.*[A-Z].*[A-Z])        Ensure string has two uppercase letters.
(?=.*[\.!@#$&*])            Ensure string has one special case letter.
(?=.*[0-9].*[0-9])        Ensure string has two digits.
(?=.*[a-z].*[a-z]) Ensure string has two lowercase letters.
.{8}                      Ensure string is of length 8.
$                         End anchor.

```
passwordPolicy = new RegExp("^(?=.*[A-Z].*[A-Z])(?=.*[!@#$&*])(?=.*[0-9].*[0-9])(?=.*[a-z].*[a-z]).{30,}$")
if (!passwordPolicy.test(password)) {
    console.warn(...)
    // bye bye baby
}
```

jkoenig134 commented:
> I'm absolutely no fan of introducing an external password-check library for such a low-level check

That lib was well maintained and had more features than a simple regex, but I removed it.

I added some more special characters to make the feature way more usable.

britsta commented on README_dev.md:
```diff
@@ -149,7 +149,7 @@ npm run test:local -- testSuiteName
       "database": { "driver": "lokijs", "folder": "./" },
       "logging": { "categories": { "default": { "appenders": ["console"] } } },
-      "infrastructure": { "httpServer": { "apiKey": "xxx", "port": 8080 } },
+      "infrastructure": { "httpServer": { "apiKey": "", "port": 8080 } },
```
Maybe use `"..."` instead of `""` as placeholder for the `"apiKey"` as it is done for the other fields (`"baseUrl"`, `"platformClientId"`, `"platformClientSecret"`).

jkoenig134 replied:
if it was a placeholder I would have used `<...>`

britsta replied:
> It disables the API key.

Good to know!

> if it was a placeholder I would have used `<...>`

I would have appreciated that too (even if the example unfortunately uses `"..."` as a placeholder instead of `"<...>"` so far).

jkoenig134 replied:
done.

britsta commented on README_dev.md:
Should the placeholder for the `apiKey` in the README.md adjusted as well from `"<api-key>"` to `"<api-key-or-empty-string>"`?

jkoenig134 replied:
There is no apiKey in README.md?

jkoenig134 replied:
ah, you mean for helm. NO, we don't want to have no api key for helm deployments.

britsta replied:
> ah, you mean for helm. NO, we don't want to have no api key for helm deployments.

Yeah, I meant helm :D Alright!

stnmtz commented:
LGTM
